### PR TITLE
refactor: updated devDependencies regarding webpack / to webpack version 5 & some cleanup

### PR DESCRIPTION
**Pull request by mfranzke**
Updated webpack to version 5, and all it's related dependencies

As I'm not that kind of a webpack user, I would probably link some of the necessary code changes to some older commits with which the code was introduced.

#### Simply updated
- [clean-css-loader](https://github.com/retyui/clean-css-loader/blob/master/CHANGELOG.md)
  - [x] major release update relates to [dropping webpack@1 support](https://github.com/retyui/clean-css-loader/blob/master/CHANGELOG.md#400)
- [html-webpack-plugin](https://github.com/jantimon/html-webpack-plugin/blob/main/CHANGELOG.md)
- ~~[node-sass-selector-importer](https://github.com/maoberlehner/node-sass-magic-importer/releases)~~ (got removed by #1373)
- [style-loader](https://github.com/webpack-contrib/style-loader/blob/master/CHANGELOG.md)
  - [x] major release is related to ["minimum supported `Node.js` version is `12.13.0`" and "minimum supported `webpack` version is `5.0.0`"](https://github.com/webpack-contrib/style-loader/releases/tag/v3.0.0)
- [svg-sprite-loader](https://github.com/JetBrains/svg-sprite-loader/blob/master/CHANGELOG.md)
- [svg-transform-loader](https://github.com/JetBrains/svg-mixer/blob/master/packages/svg-transform-loader/CHANGELOG.md)
- [svgo-loader](https://github.com/svg/svgo-loader/releases)
  - [x] Switched to SVGO 3, see its breaking changes: https://github.com/svg/svgo/releases/tag/v3.0.0
    - [x] Node.js 14+ is required for version
    - [x] stable package is replaced with native stable sort (required node 12+)
  - [x] Removed support of Webpack 4
- [ts-loader](https://github.com/TypeStrong/ts-loader/blob/master/CHANGELOG.md)
  - [x] major release update relates to ["minimum webpack version: 5" and "minimum node version: 12"](https://github.com/TypeStrong/ts-loader/blob/main/CHANGELOG.md#v900)
- ~~[no-emit-webpack-plugin](https://github.com/xipasduarte/no-emit-webpack-plugin/releases)~~ (this got removed by #1349)
- [babel-loader](https://github.com/babel/babel-loader/releases)
  - [x] major update is related to [Requiring babel ^7.12.0 and Node.js >= 14.15.0](https://github.com/babel/babel-loader/releases/tag/v9.0.0)
- [css-loader](https://github.com/webpack-contrib/css-loader/blob/master/CHANGELOG.md)
  - [x] major release update relates to ["minimum required Node.js version is 10.13.0", "minimum required webpack version is 4.27.0"](https://github.com/webpack-contrib/css-loader/blob/master/CHANGELOG.md#400-2020-07-25) and ["minimum supported Node.js version is 12.13.0", "minimum supported webpack version is 5"](https://github.com/webpack-contrib/css-loader/blob/master/CHANGELOG.md#600-2021-07-14)
- [autoprefixer](https://github.com/postcss/autoprefixer/blob/main/CHANGELOG.md)
- [webpackbar](https://github.com/unjs/webpackbar/blob/master/CHANGELOG.md)
- [webpack-hot-middleware](https://www.npmjs.com/package/webpack-hot-middleware)
  - [x] [mainly about webpack 5 support](https://github.com/webpack-contrib/webpack-hot-middleware/blob/master/CHANGELOG.md)
- [sass-loader](https://github.com/webpack-contrib/sass-loader/blob/master/CHANGELOG.md)
  - [x] BREAKING CHANGES, not relevant for us (we're updating `webpack` anyhow with this ticket and are on `node` version `14.21`):
    - minimum supported webpack version is 5
    - [minimum supported Node.js version is 14.15.0](https://github.com/webpack-contrib/sass-loader/blob/master/CHANGELOG.md)

#### Updated and migrated some code
- [webpack](https://github.com/webpack/webpack/releases): https://webpack.js.org/migrate/5/#upgrade-webpack-to-5
  - [x] ValidationError: Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema. - options has an unknown property 'sourceMap'.
  - [x] optimization.splitChunks.cacheGroups.vendors → optimization.splitChunks.cacheGroups.defaultVendors (https://webpack.js.org/migrate/5/#clean-up-configuration)
  - [x] When using a custom configuration, drop name: false and replace name: string | function with idHint: string | function (https://webpack.js.org/migrate/5/#clean-up-configuration)
  - [x] out of the following two changes, that lead to conflicting configurations after migration, I've decided to use the latter – both have been introduced with commit https://github.com/pattern-lab/patternlab-node/commit/70fe9b63676e21439ba77d5ddbdbc5b243ba829f#diff-8c9b7dba764c6bd34aaec36a6a42e7b69d9a46b19712a47706355bb8616d3319R212, but don't seem to work in parallel anymore:
    - [x] optimization.namedChunks: true → [optimization.chunkIds](https://webpack.js.org/configuration/optimization/#optimizationchunkids): 'named' (https://webpack.js.org/migrate/5/#update-outdated-options)
    - [x] optimization.occurrenceOrder: true → optimization: { chunkIds: 'total-size', [moduleIds](https://webpack.js.org/configuration/optimization/#optimizationmoduleids): 'size' } (https://webpack.js.org/migrate/5/#update-outdated-options)
- [clean-webpack-plugin](https://github.com/johnagan/clean-webpack-plugin/releases):
  - [x] https://github.com/johnagan/clean-webpack-plugin/issues/106
    - [x] "`paths`: Use `cleanOnceBeforeBuildPatterns` and/or `cleanAfterEveryBuildPatterns`."
    - [x] "`allowExternal`: Renamed to `dangerouslyAllowCleanPatternsOutsideProject`. You most likely do not need this option with version `v2` because it checks against `process.cwd()` rather than plugin option root."
  - [x] besides that no relevant breaking changes (["Dropped support for NodeJS 8" and "Dropped support for webpack 3"](https://github.com/johnagan/clean-webpack-plugin/releases/tag/v4.0.0))
- [webpack-merge](https://github.com/survivejs/webpack-merge/blob/develop/CHANGELOG.md)
  - [x] [latest release related to webpack 5 compatibility](https://github.com/survivejs/webpack-merge/blob/develop/CHANGELOG.md#580--2021-06-07)
- [mini-css-extract-plugin](https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/CHANGELOG.md)
  - [x] the major release update is mainly about ["minimum supported Node.js version is 12.13.0" and "minimum supported webpack version is 5.0.0"](https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/CHANGELOG.md#-breaking-changes)
- [webpack-cli](https://github.com/webpack/webpack-cli/blob/master/CHANGELOG.md):
  - [x] the [display](https://github.com/pattern-lab/patternlab-node/pull/1301/files#diff-a17e058b928535d4ad4fb906dfc6eb13a744a6e231e4a2fd4368c109daf64871L10) parameter seems to have been replaced by [`stats`](https://webpack.js.org/configuration/stats/#stats-presets)
  - [x] besides that it's mainly about [webpack 5 support being dropped](https://github.com/webpack/webpack-cli/blob/master/CHANGELOG.md#breaking-changes)
- [copy-webpack-plugin](https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/CHANGELOG.md):
  - [x] the flatten option has been differently defined now (https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/CHANGELOG.md#700-2020-12-10 and https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/CHANGELOG.md#800-2021-03-04)
  - [x] besides that nothing relevant for us:
    - [minimum supported `Node.js` version is `12.13.0`](https://github.com/webpack-contrib/copy-webpack-plugin/releases/tag/v9.0.0)
    - [minimum supported `Node.js` version is `12.20.0`](https://github.com/webpack-contrib/copy-webpack-plugin/releases/tag/v10.0.0)
    - [minimum supported `Node.js` version is `14.15.0`](https://github.com/webpack-contrib/copy-webpack-plugin/releases/tag/v11.0.0)
- [webpack-dev-middleware](https://github.com/webpack/webpack-dev-middleware/blob/master/CHANGELOG.md):
  - [x] "the logger, logLevel and logTime options were removed without replacement." source: https://github.com/webpack/webpack-dev-middleware/releases?after=v4.0.0-rc.1
  - [x] besides that it's mainly about supported Node.js and webpack version updates (minimum supported Node.js version is 12.13.0 (https://github.com/webpack/webpack-dev-middleware/issues/928) ([4cffeff](https://github.com/webpack/webpack-dev-middleware/commit/4cffeffb5fd07ea79e5a7a5a0cdb3f08f3856c06)) and [minimum supported webpack version is 5.0.0, minimum supported Nodejs version is 14.15.0](https://github.com/webpack/webpack-dev-middleware/blob/master/CHANGELOG.md#600-2022-11-20))
- [postcss-loader](https://github.com/webpack-contrib/postcss-loader/blob/master/CHANGELOG.md#400-2020-09-07):
  - [x] "postcss was moved to `peerDependencies`, you need to install `postcss`"
  - [x] "`PostCSS` (`plugins`/`syntax`/`parser`/`stringifier`) options was moved to the `postcssOptions` option, please look at [docs](https://github.com/webpack-contrib/postcss-loader#postcssoptions)"
  - [x] [minimum supported `Node.js` version is `12.13.0`](https://github.com/webpack-contrib/postcss-loader/blob/master/CHANGELOG.md#600-2021-06-10)
  - [x] [minimum supported `Node.js` version is `14.15.0`](https://github.com/webpack-contrib/postcss-loader/blob/master/CHANGELOG.md#700-2022-05-18)
- [html-loader](https://www.npmjs.com/package/html-loader)
  - [x] main [breaking change](https://github.com/webpack-contrib/html-loader/blob/master/CHANGELOG.md#100-2020-03-19): "the `interpolate` option was removed, please consider migration on the [preprocessor](https://github.com/webpack-contrib/html-loader#preprocessor)"; this has been handled by https://github.com/pattern-lab/patternlab-node/pull/1301/commits/35e13c6c4de344c2372026357754e9c2786154f6
  - [x] furthermore all of the other used options have been (re)moved in the meantime
  - [x] and we need to ensure that the generated assets (like e.g. `<link rel="stylesheet" href="styleguide/css/pattern-lab.css" media="all" />`) aren't being followed by Webpack, as they are partly being generated on the fly or copied to their place in the structure, by the [newly supported webpack comment](https://github.com/webpack-contrib/html-loader#disable-url-resolving-using-the----webpackignore-true----comment).
  - [x] `preventAttributesEscaping` option got introduced with https://github.com/pattern-lab/patternlab-node/commit/318801a4ac10eb7263a09ab948c5cc272a922ce9, but there's no documentation out of that version that could get reproduced anymore, so it needed to get removed; the current comment within the webpack config mentions a correctly working srcdoc-attribute that got introduced by https://github.com/pattern-lab/patternlab-node/pull/1131 but that still works correctly.
- [x] [`mini-css-extract-plugin`](https://www.npmjs.com/package/mini-css-extract-plugin): Removed `allChunks` option that has been added with https://github.com/pattern-lab/patternlab-node/commit/318801a4ac10eb7263a09ab948c5cc272a922ce9#diff-8c9b7dba764c6bd34aaec36a6a42e7b69d9a46b19712a47706355bb8616d3319R154 but is now marked as not being a parameter (any more)
 
#### Removed
- [`hard-source-webpack-plugin-patch`](https://www.npmjs.com/package/hard-source-webpack-plugin-patch): Webpack 5 supportes module caching out of the box, compare to e.g. https://github.com/mzgoddard/hard-source-webpack-plugin/issues/514#issuecomment-687621600 – and we don't need to configure anything for this:
> Cache the generated webpack modules and chunks to improve build speed. `cache` is set to `type: 'memory'` in [`development` mode](https://webpack.js.org/configuration/mode/#mode-development) and disabled in [`production` mode](https://webpack.js.org/configuration/mode/#mode-production). `cache: true` is an alias to `cache: { type: 'memory' }`. To disable caching pass `false`
source: https://webpack.js.org/configuration/cache/#cache
- ~~[critical](https://www.npmjs.com/package/critical) & [critical-css-webpack-plugin](https://www.npmjs.com/package/critical-css-webpack-plugin): this seems to have been removed (intentionally?) by https://github.com/pattern-lab/patternlab-node/commit/2cb663d053e004ab17eb5af606ce289834bb89fc#diff-8c9b7dba764c6bd34aaec36a6a42e7b69d9a46b19712a47706355bb8616d3319L365~~ (these have been removed with #1319 even already)

#updatingdependencies[bit](https://github.com/pattern-lab/patternlab-node/pull/1299)by[bit](https://github.com/pattern-lab/patternlab-node/pull/1300)by[bit](https://github.com/pattern-lab/patternlab-node/pull/1302)by[bit](https://github.com/pattern-lab/patternlab-node/pull/1303)by[bit](https://github.com/pattern-lab/patternlab-node/pull/1305)by[bit](https://github.com/pattern-lab/patternlab-node/pull/1306)by[bit](https://github.com/pattern-lab/patternlab-node/pull/1307)by[bit](https://github.com/pattern-lab/patternlab-node/pull/1320)

### CODE REVIEW

File: packages/uikit-workshop/webpack.config.js [hunk 1]
```diff
@@ -349,22 +331,19 @@ module.exports = function(apiConfig) {
...
+          verbose: false,
+          cleanOnceBeforeBuildPatterns: config.watch
             ? []
             : [
                 `${config.buildDir}/index.html`,
                 `${config.buildDir}/styleguide/css`,
                 `${config.buildDir}/styleguide/js`,
               ],
-          {
-            allowExternal: true,
```
Comment (mfranzke) [inline packages/uikit-workshop/webpack.config.js:355]: @sghoweri Do we really (still) need this? Introduced with https://github.com/pattern-lab/patternlab-node/blob/3d38d4399bc8553b68bf5914fc94af3351bb49d6/packages/uikit-workshop/webpack.config.js#L213 - `allowExternal` got replaced by `dangerouslyAllowCleanPatternsOutsideProject` with a newer version of the plugin, see https://github.com/johnagan/clean-webpack-plugin/issues/106, but question is whether we need it at all (any more). Probably it was meant to allow cleaning the folder `dist/` as the `output.path` is defined as `dist/styleguide`

File: packages/uikit-workshop/webpack.config.js [hunk 2]
```diff
@@ -1,10 +1,8 @@
 // webpack.config.js
-const CleanWebpackPlugin = require('clean-webpack-plugin');
-const HardSourceWebpackPlugin = require('hard-source-webpack-plugin-patch');
+const { CleanWebpackPlugin } = require('clean-webpack-plugin');
 const TerserPlugin = require('terser-webpack-plugin');
 const NoEmitPlugin = require('no-emit-webpack-plugin');
 const autoprefixer = require('autoprefixer');
-const CriticalCssPlugin = require('critical-css-webpack-plugin');
```
Comment (mfranzke) [inline packages/uikit-workshop/webpack.config.js]: @sghoweri: this seems to have been removed (intentionally?) by https://github.com/pattern-lab/patternlab-node/commit/2cb663d053e004ab17eb5af606ce289834bb89fc#diff-8c9b7dba764c6bd34aaec36a6a42e7b69d9a46b19712a47706355bb8616d3319L365 - is that correct ? Then I'd like to remove the module now

File: packages/uikit-workshop/webpack.config.js [hunk 3]
```diff
@@ -373,7 +350,6 @@ module.exports = function(apiConfig) {
         new MiniCssExtractPlugin({
           filename: `[name].css`,
           chunkFilename: `[id].css`,
-          allChunks: true,
```
Comment (mfranzke) [inline packages/uikit-workshop/webpack.config.js:370]: @sghoweri could you please evaluate whether it's (still) necessary? it has been added with https://github.com/pattern-lab/patternlab-node/commit/318801a4ac10eb7263a09ab948c5cc272a922ce9#diff-8c9b7dba764c6bd34aaec36a6a42e7b69d9a46b19712a47706355bb8616d3319R154 but is now marked as not being a parameter (any more)

File: packages/uikit-workshop/webpack.config.js [hunk 4]
```diff
@@ -309,7 +290,6 @@ module.exports = function(apiConfig) {
...
```
Comment (mfranzke) [inline packages/uikit-workshop/webpack.config.js:306]: @sghoweri could you please evaluate whether it's (still) necessary? it has been added with https://github.com/pattern-lab/patternlab-node/commit/318801a4ac10eb7263a09ab948c5cc272a922ce9#diff-8c9b7dba764c6bd34aaec36a6a42e7b69d9a46b19712a47706355bb8616d3319R154 but is now marked as not being a parameter (any more)

File: packages/uikit-workshop/webpack.config.js [hunk 5]
```diff
@@ -260,27 +257,11 @@ module.exports = function(apiConfig) {
       plugins: [
         new WebpackBar(),
         new CopyPlugin(config.copy),
+        // TODO: Check whether this is still correctly implemented (correct filename?), introduced with https://github.com/pattern-lab/patternlab-node/commit/318801a4ac10eb7263a09ab948c5cc272a922ce9#diff-8c9b7dba764c6bd34aaec36a6a42e7b69d9a46b19712a47706355bb8616d3319R156 - especially if this is meant to work in combination with Mini CSS Extract Plugin it's probably a little simpler to declare now: https://github.com/xipasduarte/no-emit-webpack-plugin/releases/tag/v4.0.1
         new NoEmitPlugin(['css/pattern-lab.js']),
```
Comment (mfranzke) [inline packages/uikit-workshop/webpack.config.js]: @sghoweri could you please check whether this is still necessary / correctly implemented (correct filename?); the code was introduced with https://github.com/pattern-lab/patternlab-node/commit/318801a4ac10eb7263a09ab948c5cc272a922ce9#diff-8c9b7dba764c6bd34aaec36a6a42e7b69d9a46b19712a47706355bb8616d3319R156 - especially if this is meant to work in combination with Mini CSS Extract Plugin it's probably a little simpler to declare now: https://github.com/xipasduarte/no-emit-webpack-plugin/releases/tag/v4.0.1

File: packages/uikit-workshop/webpack.config.js (hunk 1 above)
Comment (JosefBredereck) [reply packages/uikit-workshop/webpack.config.js:355]: If the build is still ok and the preview loads up correctly after building the ui-kit I would assume we do not need the parameter anymore.

File: packages/uikit-workshop/webpack.config.js (hunk 2 above)
Comment (JosefBredereck) [reply packages/uikit-workshop/webpack.config.js]: Please proceed by removing it 👍

File: packages/uikit-workshop/webpack.config.js (hunk 3 above)
Comment (JosefBredereck) [reply packages/uikit-workshop/webpack.config.js:370]: If the build is still ok and the preview loads up correctly after building the ui-kit I would assume we do not need the parameter anymore.

File: packages/uikit-workshop/webpack.config.js (hunk 4 above)
Comment (JosefBredereck) [reply packages/uikit-workshop/webpack.config.js:306]: If the build is still ok and the preview loads up correctly after building the ui-kit I would assume we do not need the parameter anymore.

**Comment by JosefBredereck** [comment]
What is the current status of this PR? Do we want to proceed with it?

**Comment by mfranzke** [comment]
> What is the current status of this PR? Do we want to proceed with it?

I'm sadly not focussing on one activity at the time and jumping from topic to topic – I would say that this is 95% done, but haven't had the chance to finalize it. Looks promising at least.

File: packages/uikit-workshop/webpack.config.js (hunk 3 above)
Comment (mfranzke) [reply packages/uikit-workshop/webpack.config.js:370]: I've tested it, it's still okay.

File: packages/uikit-workshop/webpack.config.js (hunk 4 above)
Comment (mfranzke) [reply packages/uikit-workshop/webpack.config.js:306]: I've tested it, it's still okay.

**Comment by mfranzke** [comment]
@JosefBredereck I've finally checked this and it looks good to me for approval and merge.

**Pull request merged**